library/std/src/io/error.rs

```diff
@@ -465,6 +466,19 @@ impl From<ErrorKind> for Error {
     }
 }
 
+#[stable(feature = "io_error_from_try_reserve", since = "1.62.0")]
+impl From<TryReserveError> for Error {
+    #[inline]
+    fn from(error: TryReserveError) -> Error {
+        let kind = match error.kind() {
+            TryReserveErrorKind::CapacityOverflow => ErrorKind::InvalidInput,
+            TryReserveErrorKind::AllocError { .. } => ErrorKind::OutOfMemory,
+        };
+
+        Error::new(kind, error)
```

Comment:
Note that this involves a double allocation, effectively a `Box<Box<dyn io::Error>>`.

Reply:
I didn't realize `io::Error::new` was a double-allocation! I wonder if [thin dyn box](https://github.com/rust-lang/rust/issues/92791) would solve that.

Reply:
Even worse: if you pass it a string literal (`
Error::new(kind, "...")`) you get a _triple_ allocation (and a strcpy/memcpy) as the inner box will contain a `String`.

ThinBox can save one allocation here. I think that'd be a simple change to make, but there might be some non-obvious pitfall.

Reply:
See https://github.com/rust-lang/rust/issues/83352